File: src/controllers/recurringtransaction.ts
/**
 * Handles recurring transaction data interface
 */

import { Records, Record, RecordMap, firebase } from "./records";

interface RecurringTransaction extends Record {
    amount : number;
    cash? : boolean;
    category : string;
    end : string;
    name : string;
    note? : string;
    period : string;
    start : string;
    transfer? : boolean;
}

export default class RecurringTransactions extends Records<RecurringTransaction> {
    constructor(reference: firebase.database.Reference) {
        super(reference); 
    }

    sanitizeAfterRead(transaction: RecurringTransaction) : RecurringTransaction {
        transaction.cash = transaction.cash || false;
        transaction.note = transaction.note || null;
        transaction.transfer = transaction.transfer || false;

        return transaction;
    }

    sanitizeBeforeWrite(transaction: RecurringTransaction) : RecurringTransaction {
        transaction.cash = transaction.cash? transaction.cash : null;
        transaction.note = transaction.note === undefined || transaction.note === "" ? null : transaction.note;
        transaction.transfer = transaction.transfer? transaction.transfer : null;

        return transaction;
    }
}

File: iosapp/models/recurringtransaction.swift
import Firebase

class RecurringTransaction: Record {
    var amount: Double = 0.0
    var cash: Bool
    var category: String
    var end: String
    var name: String
    var note: String?
    var period: String
    var start: String
    var transfer: Bool

    var deposit : Bool {
        get { return value >= 0 }
    }

    init(snapshot: DatabaseSnapshot) {
        super(snapshot)
        if !self.fromObject(snapshot.value) return nil
    }

    init(data: [AnyHashable: Any]) {
        super(data)
        if !self.fromObject(data) return nil
    }

    private func fromObject(value: [AnyHashable: Any]) -> Bool {
        guard 
            let amount = value["amount"] as? Double,
            let category = value["category"] as? String,
            let end = value["end"] as? String,
            let name = value["name"] as? String,
            let period = value["period"] as? String,
            let start = value["start"] as? String
        else {
            return false
        }

        self.amount = amount
        self.category = category
        self.end = end
        self.name = name
        self.period = period
        self.start = start

        self.cash = value["cash"] as? Bool ?? false
        self.note = value["note"] as? String
        self.transfer = value["transfer"] as? Bool ?? false

        return true
    }

    func asObject() -> [AnyHashable: Any] {
        var result = [:]

        result["amount"] = self.amount
        result["category"] = self.category
        result["end"] = self.end
        result["name"] = self.name
        result["period"] = self.period
        result["start"] = self.start

        // nil values are ignored on the firebase end, just set them
        result["check"] = self.check
        result["note"] = self.note
        result["recurring"] = self.recurring

        // avoid storing false boolean values since nils are also parsed
        if self.cash { result["cash"] = true }
        if self.paid { result["paid"] = true }
        if self.transfer { result["transfer"] = true }

        return result
    }
}